tests/Unit/GuzzleFileFetcherTest.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Tuf\Tests\Unit;
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Handler\MockHandler;
+use GuzzleHttp\HandlerStack;
+use GuzzleHttp\Middleware;
+use GuzzleHttp\Psr7\Response;
+use PHPUnit\Framework\TestCase;
+use Tuf\Client\GuzzleFileFetcher;
+use Tuf\Exception\DownloadSizeException;
+use Tuf\Exception\RepoFileNotFound;
+
+/**
+ * @coversDefaultClass \Tuf\Client\GuzzleFileFetcher
+ */
+class GuzzleFileFetcherTest extends TestCase
+{
+    /**
+     * Data provider for testFetchFile().
+     *
+     * @return array[]
+     *   Sets of arguments to pass to the test method.
+     */
+    public function dataProvider(): array
+    {
+        return [
+            [404, 256, RepoFileNotFound::class],
+            [403, 256, 'RuntimeException'],
+            [500, 256, 'RuntimeException'],
+            [200, 4, DownloadSizeException::class],
+            [200, 256, null],
+        ];
+    }
+
+    /**
+     * Tests various conditions when fetching a file.
+     *
+     * @param integer $statusCode
+     *   The response status code.
+     * @param integer $maxBytes
+     *   The maximum number of bytes to read from the response.
+     * @param string|null $exceptionClass
+     *   The expected exception class that will be thrown, if we expect the file
+     *   fetcher to throw an exception.
+     *
+     * @return void
+     *
+     * @dataProvider dataProvider
+     *
+     * @covers ::fetchFile
+     */
+    public function testFetchFile(int $statusCode, int $maxBytes, ?string $exceptionClass): void
+    {
+        $content = json_encode(['oolong', 'assam', 'matcha', 'herbal']);
+        $mockHandler = new MockHandler();
+        $mockHandler->append(new Response($statusCode, [], $content));
+        $handlerStack = HandlerStack::create($mockHandler);
+        // The httpErrors() middleware will throw an exception if the status
+        // code is not 200.
+        $handlerStack->push(Middleware::httpErrors());
+        $history = [];
+        $handlerStack->push(Middleware::history($history));
+        $client = new Client(['handler' => $handlerStack]);
+
+        $fetcher = new GuzzleFileFetcher($client);
+        if ($exceptionClass) {
+            $this->expectException($exceptionClass);
+        }
+        $data = $fetcher->fetchFile('test.json', $maxBytes);
+        $this->assertSame($content, $data);
```

Comment:
This seems like it could be just 2 test methods. For all the cases where there is exception the last assert is never called. So maybe one for requests with exceptions and 1 without.

everything above this segment code be refactor into a common function. Like `\Tuf\Tests\Unit\Client\UpdaterTest::getSystemInTest()` because are using that pattern multiple places in our tests

Reply:
> This seems like it could be just 2 test methods. For all the cases where there is exception the last assert is never called.

For the record, that was totally intentional on my part :) But I agree with you that it's kinda implict non-obvious programming, so I'm okay with splitting this into two methods.

Reply:
obvious seems better. So lets split it